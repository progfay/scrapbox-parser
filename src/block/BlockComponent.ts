export type BlockComponentType = {
  readonly indent: number
  readonly text: string
}

export const convertToBlockComponent = (block: string): BlockComponentType => ({
  indent: block.match(/^\s+/)?.[0].length ?? 0,
  text: block
})

export const convertToBlockComponents = (blocks: string): BlockComponentType[] => (
  blocks.split('\n').map(convertToBlockComponent)
)
