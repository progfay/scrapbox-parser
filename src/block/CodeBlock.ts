import type { Row } from "./Row";

export interface CodeBlockPack {
	type: "codeBlock";
	rows: Row[];
}

/**
 * Scrapbox {@link https://scrapbox.io/help/Syntax#58348ae2651ee500008d67df | code block} type
 */
export interface CodeBlock {
	indent: number;
	type: "codeBlock";
	fileName: string;
	content: string;
}

export const convertToCodeBlock = (pack: CodeBlockPack): CodeBlock => {
	const {
		rows: [head, ...body],
	} = pack;
	const { indent = 0, text = "" } = head ?? {};
	const fileName: string = text.replace(/^\s*code:/, "");

	return {
		indent,
		type: "codeBlock",
		fileName,
		content: body
			.map((row: Row): string => row.text.substring(indent + 1))
			.join("\n"),
	};
};
