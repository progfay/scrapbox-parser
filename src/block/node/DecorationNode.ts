import { NodeParserType, LineNodeType, convertToLineNodes } from '.'

const decorationRegExp = /^(.*?)\[([!"#%&'()*+,-./{|}<>_~]+) ((?:\[[^\]]+\]|[^\]])+)\](.*)$/

export type DecorationCharType = '*' | '!' | '"' | '#' | '%' | '&' | '\'' | '(' | ')' | '+' | ',' | '-' | '.' | '/' | '{' | '|' | '}' | '<' | '>' | '_' | '~'
export type AsteriskDecorationCharType = '*-1' | '*-2' | '*-3' | '*-4' | '*-5' | '*-6' | '*-7' | '*-8' | '*-9' | '*-10'
export type DecorationType = Exclude<DecorationCharType, '*'> | AsteriskDecorationCharType
export type DecorationNodeType = {
  readonly type: 'decoration'
  readonly decos: ReadonlyArray<DecorationType>
  readonly nodes: ReadonlyArray<LineNodeType>
}

const createDecorationNode = (decoChars: string, nodes: LineNodeType[]): DecorationNodeType => {
  const decoSet = new Set<string>(decoChars)
  if (decoSet.has('*')) {
    const asteriskCount = decoChars.split('*').length - 1
    decoSet.delete('*')
    decoSet.add(`*-${Math.min(asteriskCount, 10)}` as DecorationCharType)
  }

  return {
    type: 'decoration',
    decos: Array.from(decoSet) as DecorationType[],
    nodes
  }
}

export const DecorationNodeParser: NodeParserType = (text, { nested, quoted }, next) => {
  if (nested) return next()

  const decorationMatch = text.match(decorationRegExp)
  if (!decorationMatch) return next()

  const [, left, decoChars, target, right] = decorationMatch
  return [
    ...convertToLineNodes(left, { nested, quoted }),
    createDecorationNode(decoChars, convertToLineNodes(target, { nested: true, quoted })),
    ...convertToLineNodes(right, { nested, quoted })
  ]
}
