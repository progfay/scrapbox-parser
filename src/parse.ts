import { BlockType, convertToBlock } from './block'
import { BlockComponentType, convertToBlockComponents } from './block/BlockComponent'
import { PackedBlockComponentType, packBlockComponents } from './block/PackedBlockComponent'

type PageType = {
  title: string
  blocks: BlockType[]
}

export const convertToBlocks = (blockComponents: BlockComponentType[]): BlockType[] => {
  const packedBlockComponents: PackedBlockComponentType[] = packBlockComponents(blockComponents)
  return packedBlockComponents.map(convertToBlock)
}

export const parse = (input: string): PageType => {
  const blockComponents: BlockComponentType[] = convertToBlockComponents(input.trim())

  const [firstBlock, ...body] = blockComponents
  const title = firstBlock && firstBlock.text ? firstBlock.text : 'Untitled'
  const blocks = convertToBlocks(body)

  return { title, blocks }
}
