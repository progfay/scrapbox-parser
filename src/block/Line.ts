import { BlockComponentType } from './BlockComponent'
import { LineNodeType, convertToLineNodes } from './node'

export type LineComponentType = {
  readonly type: 'line'
  readonly component: BlockComponentType
}

export type LineType = {
  readonly indent: number
  readonly type: 'line'
  readonly nodes: ReadonlyArray<LineNodeType>
}

export const convertToLine = (lineComponent: LineComponentType): LineType => {
  const { indent, text } = lineComponent.component
  return {
    indent,
    type: 'line',
    nodes: convertToLineNodes(text.substring(indent))
  }
}
