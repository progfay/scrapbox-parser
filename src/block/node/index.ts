import { QuoteNodeParser } from './QuoteNode'
import { HelpfeelNodeParser } from './HelpfeelNode'
import { StrongImageNodeParser } from './StrongImageNode'
import { StrongIconNodeParser } from './StrongIconNode'
import { StrongNodeParser } from './StrongNode'
import { FormulaNodeParser } from './FormulaNode'
import { DecorationNodeParser } from './DecorationNode'
import { CodeNodeParser } from './CodeNode'
import { CommandLineNodeParser } from './CommandLineNode'
import { BlankNodeParser } from './BlankNode'
import { ImageNodeParser } from './ImageNode'
import { ExternalLinkNodeParser } from './ExternalLinkNode'
import { GoogleMapNodeParser } from './GoogleMapNode'
import { InternalLinkNodeParser } from './InternalLinkNode'
import { IconNodeParser } from './IconNode'
import { HashTagNodeParser } from './HashTagNode'
import { PlainNodeParser } from './PlainNode'

import type { Node } from './type'

export interface NodeParserOption {
  nested: boolean
  quoted: boolean
}
export type NextNodeParser = () => Node[]
export type NodeParser = (text: string, opts: NodeParserOption, next?: NextNodeParser) => Node[]

const FalsyEliminator: NodeParser = (text, _, next) => {
  if (text === '') return []
  return next?.() ?? []
}

const combineNodeParsers = (...parsers: NodeParser[]) => (
  text: string = '',
  opts: NodeParserOption = { nested: false, quoted: false }
): Node[] =>
  parsers.reduceRight(
    (acc: NextNodeParser, parser: NodeParser): NextNodeParser => () => parser(text, opts, acc),
    () => PlainNodeParser(text, opts)
  )()

export const convertToNodes = combineNodeParsers(
  FalsyEliminator,
  QuoteNodeParser,
  HelpfeelNodeParser,
  CodeNodeParser,
  CommandLineNodeParser,
  FormulaNodeParser,
  BlankNodeParser,
  DecorationNodeParser,
  StrongImageNodeParser,
  StrongIconNodeParser,
  StrongNodeParser,
  ImageNodeParser,
  ExternalLinkNodeParser,
  IconNodeParser,
  GoogleMapNodeParser,
  InternalLinkNodeParser,
  HashTagNodeParser
)
