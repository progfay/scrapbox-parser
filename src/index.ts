import { parse } from './parse'

export default parse
export { BlockType } from './block'
export { CodeBlockType } from './block/CodeBlock'
export { TableType } from './block/Table'
export { LineType } from './block/Line'
export { LineNodeType } from './block/node'
export { QuoteNodeType } from './block/node/QuoteNode'
export { HelpfeelNodeType } from './block/node/HelpfeelNode'
export { StrongImageNodeType } from './block/node/StrongImageNode'
export { StrongIconNodeType } from './block/node/StrongIconNode'
export { StrongNodeType } from './block/node/StrongNode'
export { FormulaNodeType } from './block/node/FormulaNode'
export { DecorationNodeType, DecorationType } from './block/node/DecorationNode'
export { CodeNodeType } from './block/node/CodeNode'
export { BlankNodeType } from './block/node/BlankNode'
export { UrlNodeType } from './block/node/UrlNode'
export { GoogleMapNodeType } from './block/node/GoogleMapNode'
export { InternalLinkNodeType } from './block/node/InternalLinkNode'
export { IconNodeType } from './block/node/IconNode'
export { HashTagNodeType } from './block/node/HashTagNode'
export { PlainNodeType } from './block/node/PlainNode'
