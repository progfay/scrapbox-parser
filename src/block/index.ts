import { convertToCodeBlock } from "./CodeBlock";
import { convertToLine } from "./Line";
import { convertToTable } from "./Table";
import { convertToTitle } from "./Title";

import type { CodeBlock } from "./CodeBlock";
import type { Line } from "./Line";
import type { Pack } from "./Pack";
import type { Table } from "./Table";
import type { Title } from "./Title";

/**
 * Scrapbox block type
 */
export type Block = Title | CodeBlock | Table | Line;

export const convertToBlock = (pack: Pack): Block => {
	switch (pack.type) {
		case "title":
			return convertToTitle(pack);

		case "codeBlock":
			return convertToCodeBlock(pack);

		case "table":
			return convertToTable(pack);

		case "line":
			return convertToLine(pack);
	}
};
