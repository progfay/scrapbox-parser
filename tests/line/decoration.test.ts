/* global describe it expect */

import { BlockComponentType, convertToBlockComponents } from '../../src/block/BlockComponent'
import { BlockType } from '../../src/block'
import { LineType } from '../../src/block/Line'
import { DecorationType, DecorationNodeType } from '../../src/block/node/DecorationNode'
import { convertToBlocks } from '../../src/parse'

const createDecorartionNode = (decos: DecorationType[], text: string): LineType => ({
  indent: 0,
  type: 'line',
  nodes: [
    {
      type: 'decoration',
      decos,
      nodes: [
        {
          type: 'plain',
          text
        }
      ]
    }
  ]
})

describe('decoration', () => {
  it('Simple decoration', () => {
    const input = `[* deco]
[** deco]
[*** deco]
[**** deco]
[***** deco]
[****** deco]
[******* deco]
[******** deco]
[********* deco]
[********** deco]
[! deco]
[" deco]
[# deco]
[% deco]
[& deco]
[' deco]
[( deco]
[) deco]
[+ deco]
[, deco]
[- deco]
[. deco]
[/ deco]
[{ deco]
[| deco]
[} deco]
[< deco]
[> deco]
[_ deco]
[~ deco]`
    const decosArray: DecorationType[][] = [
      ['*-1'], ['*-2'], ['*-3'], ['*-4'], ['*-5'], ['*-6'], ['*-7'], ['*-8'], ['*-9'], ['*-10'],
      ['!'], ['"'], ['#'], ['%'], ['&'], ['\''], ['('], [')'], ['+'], [','], ['-'], ['.'], ['/'], ['{'], ['|'], ['}'], ['<'], ['>'], ['_'], ['~']
    ]
    const expected: BlockType[] = decosArray.map((decos: DecorationType[]): LineType => createDecorartionNode(decos, 'deco'))
    expect(input).toEqualWhenParsing(expected)
  })

  it('All decoration', () => {
    const input = '[**********!"#%&\'()*+,-./{|}<>_~ decos]'
    const blockComponents: BlockComponentType[] = convertToBlockComponents(input)
    const blocks: BlockType[] = convertToBlocks(blockComponents)
    const received = ((blocks[0] as LineType).nodes[0] as DecorationNodeType).decos
    const decos: DecorationType[] = ['*-10', '!', '"', '#', '%', '&', '\'', '(', ')', '+', ',', '-', '.', '/', '{', '|', '}', '<', '>', '_', '~']
    expect(new Set<DecorationType>(received)).toEqual(new Set<DecorationType>(decos))
  })

  it('Decoration * overflow', () => {
    const decos: DecorationType[] = ['*-10']
    const expected: LineType[] = [createDecorartionNode(decos, '11*')]
    expect('[*********** 11*]').toEqualWhenParsing(expected)
  })

  it('Decoration similar with externalLink', () => {
    expect('[* hoge https://example.com]').toEqualWhenParsing([
      {
        indent: 0,
        type: 'line',
        nodes: [
          {
            type: 'decoration',
            decos: ['*-1'],
            nodes: [
              {
                type: 'plain',
                text: 'hoge '
              },
              {
                type: 'link',
                pathType: 'absolute',
                href: 'https://example.com',
                content: ''
              }
            ]
          }
        ]
      }
    ])
  })
})
