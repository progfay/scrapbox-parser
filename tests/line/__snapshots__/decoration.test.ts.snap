// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decoration Decoration * overflow 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-10",
        ],
        "nodes": [
          {
            "raw": "11*",
            "text": "11*",
            "type": "plain",
          },
        ],
        "raw": "[*********** 11*]",
        "rawDecos": "***********",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
]
`;

exports[`decoration Decoration similar with externalLink 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "raw": "hoge ",
            "text": "hoge ",
            "type": "plain",
          },
          {
            "content": "",
            "href": "https://example.com",
            "pathType": "absolute",
            "raw": "https://example.com",
            "type": "link",
          },
        ],
        "raw": "[* hoge https://example.com]",
        "rawDecos": "*",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
]
`;

exports[`decoration Decoration with hashTag 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "href": "tag",
            "raw": "#tag",
            "type": "hashTag",
          },
        ],
        "raw": "[* #tag]",
        "rawDecos": "*",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
]
`;

exports[`decoration Decoration with icon notation 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "path": "progfay",
            "pathType": "relative",
            "raw": "[progfay.icon]",
            "type": "icon",
          },
        ],
        "raw": "[* [progfay.icon]]",
        "rawDecos": "*",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
]
`;

exports[`decoration Decoration with many [ 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "!",
        ],
        "nodes": [
          {
            "raw": "[[[[[[a",
            "text": "[[[[[[a",
            "type": "plain",
          },
        ],
        "raw": "[! [[[[[[a]",
        "rawDecos": "!",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
]
`;

exports[`decoration Decoration with many [ and link 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "!",
        ],
        "nodes": [
          {
            "raw": "[[[[[",
            "text": "[[[[[",
            "type": "plain",
          },
          {
            "content": "",
            "href": "a",
            "pathType": "relative",
            "raw": "[a]",
            "type": "link",
          },
        ],
        "raw": "[! [[[[[[a]]",
        "rawDecos": "!",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
]
`;

exports[`decoration Decoration with strong icon notation (it's just icon, not strong) 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "raw": "[",
            "text": "[",
            "type": "plain",
          },
          {
            "path": "progfay",
            "pathType": "relative",
            "raw": "[progfay.icon]",
            "type": "icon",
          },
        ],
        "raw": "[* [[progfay.icon]]",
        "rawDecos": "*",
        "type": "decoration",
      },
      {
        "raw": "]",
        "text": "]",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`decoration Decoration with strong image notation (it's just image, not strong) 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "raw": "[",
            "text": "[",
            "type": "plain",
          },
          {
            "link": "",
            "raw": "[https://example.com/image.png]",
            "src": "https://example.com/image.png",
            "type": "image",
          },
        ],
        "raw": "[* [[https://example.com/image.png]]",
        "rawDecos": "*",
        "type": "decoration",
      },
      {
        "raw": "]",
        "text": "]",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`decoration Decoration with strong notation (it's just link) 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "raw": "[",
            "text": "[",
            "type": "plain",
          },
          {
            "content": "",
            "href": "link",
            "pathType": "relative",
            "raw": "[link]",
            "type": "link",
          },
        ],
        "raw": "[* [[link]]",
        "rawDecos": "*",
        "type": "decoration",
      },
      {
        "raw": "]",
        "text": "]",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`decoration Simple decoration 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[* deco]",
        "rawDecos": "*",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-2",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[** deco]",
        "rawDecos": "**",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-3",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[*** deco]",
        "rawDecos": "***",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-4",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[**** deco]",
        "rawDecos": "****",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-5",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[***** deco]",
        "rawDecos": "*****",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-6",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[****** deco]",
        "rawDecos": "******",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-7",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[******* deco]",
        "rawDecos": "*******",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-8",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[******** deco]",
        "rawDecos": "********",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-9",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[********* deco]",
        "rawDecos": "*********",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-10",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[********** deco]",
        "rawDecos": "**********",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "!",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[! deco]",
        "rawDecos": "!",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          """,
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[" deco]",
        "rawDecos": """,
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "#",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[# deco]",
        "rawDecos": "#",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "%",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[% deco]",
        "rawDecos": "%",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "&",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[& deco]",
        "rawDecos": "&",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "'",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[' deco]",
        "rawDecos": "'",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "(",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[( deco]",
        "rawDecos": "(",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          ")",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[) deco]",
        "rawDecos": ")",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "+",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[+ deco]",
        "rawDecos": "+",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          ",",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[, deco]",
        "rawDecos": ",",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "-",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[- deco]",
        "rawDecos": "-",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          ".",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[. deco]",
        "rawDecos": ".",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "/",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[/ deco]",
        "rawDecos": "/",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "{",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[{ deco]",
        "rawDecos": "{",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "|",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[| deco]",
        "rawDecos": "|",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "}",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[} deco]",
        "rawDecos": "}",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "<",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[< deco]",
        "rawDecos": "<",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          ">",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[> deco]",
        "rawDecos": ">",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "_",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[_ deco]",
        "rawDecos": "_",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "~",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[~ deco]",
        "rawDecos": "~",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
]
`;
